Improve Mistral models integration with llama.cpp

juliendenize opened the pull request:
## Description

This PR aims to enhance the integration of Mistral models with llama.cpp by addressing several key issues and introducing new features. Here are the details:

### Context

- The current HF conversion to GGUF does not work directly for Mistral models due to our format that is vLLM based. This means that we have to first convert weights to Hugging Face then to GGUF which is not ideal and can lead to conversion errors if the first conversion is not done correctly. It also means that adding new models to the llama.cpp ecosystem requires first adding them to Transformers.
- We do not support chat templates natively which means chat templates are community based and not guaranteed to work correctly.
- We are using [mistral-common](https://github.com/mistralai/mistral-common) internally for tokenization and want the community to use it to unlock full capacities of our models. As mistral-common is a Python library, we have opened a [PR](https://github.com/mistralai/mistral-common/pull/113) to add a REST API via FastAPI to make it easier for users who are not in the Python ecosystem.

### Using mistral-common with llama.cpp

We recommend that users only use the `llama-server` tool with the `/completions` route of the server for now, as it is the only one that supports tokens input. We also advise users to set `return_tokens=True` in their requests to let `mistral-common` handle detokenization.

### Added features

1. Model conversion:

We have added a script to convert Mistral models to GGUF directly from Hugging Face. This script is located at `convert_mistral_to_gguf.py` and can be used to convert Mistral models to GGUF format.

2. Model architecture:

We registered the Mistral architecture in llama.cpp to support Mistral models natively. This allows users to use Mistral models with llama.cpp without having to convert them to Hugging Face first.

### Known Limitations:

Our approach does not support multimodality:
- mistral-common handles processing multimodal data but they cannot be passed to llama.cpp via the route.
- llama.cpp only supports multimodality via chat templates, which we do not support.

Also this approach requires users to only use the llama.cpp server with the `/completions` route.

## Example Code

To get started, install mistral-common using the following command:

#### (Optional) Convert the model

```python
HF_TOKEN=... python convert_mistral_to_gguf.py \
mistralai/Devstral-Small-2505 --remote --ctx-train 131072 --outtype bf16
```

#### Launch the mistral-common and llama.cpp servers

```bash
pip install git+https://github.com/mistralai/mistral-common.git@improve_llama_cpp_integration[server]
```

Launch the mistral-common server:

```bash
HF_TOKEN=... mistral_common mistralai/Devstral-Small-2505 --port 6000
```

Launch the llama.cpp server:

```bash
./build/bin/llama-server -m models/Devstral-Small-2505-Q4_K_M.gguf --port 8080
```

#### Use the servers

Here is a code snippet demonstrating how to use the new features:

```python
import requests

mistral_common_url = "http://127.0.0.1:6000"
llama_cpp_url = "http://127.0.0.1:8080"

def tokenize(messages, url):
    response = requests.post(f"{url}/tokenize/messages", json=messages)
    return response.json()

def detokenize(tokens, url):
    response = requests.post(f"{url}/detokenize", json={"tokens": tokens})
    return response.json()

def detokenize_message(tokens, url):
    response = requests.post(f"{url}/detokenize", json={"tokens": tokens, "as_message": True})
    return response.json()

def generate(tokens, url):
    response = requests.post(f"{url}/completions", json={
        "prompt": tokens,
        "stream": False,
        "return_tokens": True
    })
    return response.json()

messages = [
    {"role": "system", "content": "You are Devstral a cool coding agent that can help users with their coding needs."},
    {"role": "user", "content": "Who are you and what can you do?"}
]

tokens = tokenize(messages, mistral_common_url)
print(tokens)

generated = generate(tokens, llama_cpp_url)["tokens"]
print(generated)

detokenized = detokenize(generated, mistral_common_url)
print(detokenized)

detokenized_message = detokenize_message(generated, mistral_common_url)
print(detokenized_message)
```

## Feedback and Contributions

We believe these changes will significantly improve the integration of Mistral models with llama.cpp and provide a better experience for our users. We welcome any feedback or suggestions to further enhance this integration. Also, as we have few experience in the codebase of llama.cpp, we welcome any help to improve the integration and make sure we respect the codebase and the community.

juliendenize committed:
Improve Mistral models integration with llama.cpp

ggerganov commented:
Thanks for the contribution. From a developer perspective, it looks like a good approach to avoid any potential tokenization / formatting problems. In general, for all models, using a reference tokenizer instead of relying on `llama.cpp` is always recommended. From usability standpoint, the requirement to start a separate tokenization server is a bit of a drawback, but I understand that correctness is of higher importance.

My understanding is that most chat template problems occur during the early days of the model release, and with time tend to get polished and fixed. So this approach would be a stable alternative during such periods of instability.

CISC commented on convert_mistral_to_gguf.py:
```diff
@@ -0,0 +1,1118 @@
+import numpy as np
+import torch
+
+from gguf.constants import MODEL_ARCH, MODEL_ARCH_NAMES
+from gguf.vocab import MistralTokenizerType, MistralVocab
+from mistral_common.tokens.tokenizers.multimodal import DATASET_MEAN, DATASET_STD
+
+if TYPE_CHECKING:
+    from torch import Tensor
+
+if "NO_LOCAL_GGUF" not in os.environ:
+    sys.path.insert(1, str(Path(__file__).parent / "gguf-py"))
+import gguf
+
```
```suggestion
if "NO_LOCAL_GGUF" not in os.environ:
    sys.path.insert(1, str(Path(__file__).parent / "gguf-py"))
import gguf

from gguf.constants import MODEL_ARCH, MODEL_ARCH_NAMES
from gguf.vocab import MistralTokenizerType, MistralVocab
from mistral_common.tokens.tokenizers.multimodal import DATASET_MEAN, DATASET_STD

if TYPE_CHECKING:
    from torch import Tensor

```
This should hopefully also allow the CI to complete.

CISC commented:
>     * I tried to make the CI green but I think that because we modified gguf-py files some checks cannot pass because it installs the published package AFAIU. Is it right ?

Partially, there's also a `pydantic` version conflict:
https://github.com/ggml-org/llama.cpp/actions/runs/16474192826/job/46571944794?pr=14737#step:4:291

Kreijstal commented:
how to do transcription only?

juliendenize committed:
Revert changes and fix gguf

juliendenize committed:
Revert change

juliendenize commented:
> @juliendenize Please undo all the formatting/style changes, they are not relevant and add too much noise to the PR, will review afterwards. :)

Done sorry about that my own formatter was on.

Is there a formatter or linter available for Python ? Didn't find it in the contributing guidelines.
I've installed flake8 but it didn't flag any thing when launched locally

CISC commented:
Pillow conflict, should be fine to update:
https://github.com/ggml-org/llama.cpp/actions/runs/16497338367/job/46646389345?pr=14737#step:4:305

CISC commented:
Right, now we are getting somewhere. :)
https://github.com/ggml-org/llama.cpp/actions/runs/16497849941/job/46648060083?pr=14737

am17an commented:
@juliendenize do you also plan to make changes to `convert_mistral_to_gguf.py` to have mappings for `audio_tower` for the mmproj, it will be necessary for the new voxtral models I guess?

juliendenize commented:
> @juliendenize do you also plan to make changes to `convert_mistral_to_gguf.py` to have mappings for `audio_tower.*` for the mmproj, I guess it will be necessary for the new voxtral models?

that would be cool indeed, I didn't work personally on Voxtral (for the model), so I might need some assistance as I lack experience in audio models.

Is voxtral already supported by llama.cpp ? I assumed that not for now.

ngxson commented:
I've just had a deeper look into this PR. One concern though, most of the code inside `convert_mistral_to_gguf.py` is copied from `convert_hf_to_gguf.py`, which can make it a bit tricky in the long term, especially the code for multimodal model conversion.

Just thinking, maybe it's better to bring them right into `convert_hf_to_gguf.py`? AFAIU most of the complicated code in this PR is dedicated to converting the tokenizer to GGUF.

Btw, I'm also working converting Voxtral to GGUF. I thought that would be simple but I'm currently stuck at the tokenizer. Trying a quick hack to copy some code from this PR.. will see if it work.

ngxson commented:
Ok so as demo in https://github.com/ggml-org/llama.cpp/pull/14862, I think it might be better to merge everything into `convert_hf_to_gguf`. This has 2 big advantages:
- Easier for long term maintenance, since we will have less duplicated code
- Less confusion for end-users. Users who are not very familiar with llama.cpp may not understand that they need to use a dedicated script for mistral models (or models fine tuned from mistral)

juliendenize commented:
Hi @ngxson thanks for the review.

> Just thinking, maybe it's better to bring them right into convert_hf_to_gguf.py?

The reason I split the two files was to avoid confusion of what is happening, because here we don't convert hf models.

It is indeed a lot of copy paste from convert_hf_to_gguf.py but with lots of overriding. We could probably subclasses but end up overriding whole methods and use few super(). Though not entirely sure about that as I decided to decouple things really early and didn't keep track of that. I can probably achieve something better. Maybe the first thing would be to import from convert_hf_to_gguf.py . Then if you have a strong opinion about merging the two it could be done more easily.

ngxson commented:
> The reason I split the two files was to avoid confusion of what is happening, because here we don't convert hf models.

Hmm FYI, `convert_hf_to_gguf` already support some non-HF models so I think it's probably fine to merge these two.

We can also add an additional flag like `--non-hf` if needed, but eventually I think it's unnecessary.

From the perspective of `convert_hf_to_gguf`, not many things that make a mistral model different from HF model:
- The tensors are stored inside `consolidated.safetensors` --> easy to support
- `config.json` is different --> also easy to support
- Tokenizer conversion --> a bit tricky, but as demo in https://github.com/ggml-org/llama.cpp/pull/14862 it is feasible

So overall I still think merging everything into `convert_hf_to_gguf` is not that complicated and will be a better choice.

CISC commented:
> thanks @ngxson I started the process of refactoring. I have a bug that i need to fix (probably on Monday) which is why I didn't push but I prefer notifying you guys as I don't want to be silent again. You were right there is very few changes to make !
> 
> BTW @CISC we merged the PR and made the release of `mistral-common`.

Perfect, we'll resume next week then. :)

> FYI we also released Magistral GGUF https://huggingface.co/mistralai/Magistral-Small-2507-GGUF thanks to this PR seems to work very smoothly with `mistral-common` and `llama.cpp`

Nice.

BTW, it's unfortunate that you rely entirely on your own chat parsing instead of making chat templates, this forces us to use community written chat templates instead, see https://github.com/ggml-org/llama.cpp/pull/14862#discussion_r2229652708

juliendenize committed:
refactor convert_mistral_to_gguf.py in convert_hf_to_gguf.py

juliendenize committed:
Revert collateral

juliendenize committed:
Rename model name

juliendenize commented:
Hey @CISC think it should be good there I refactored what was needed.

To use the Mistral format in the conversion script I added the same flag we did for vLLM with `--mistral-format`. This avoids ambiguity with HF's implementation.

> BTW, it's unfortunate that you rely entirely on your own chat parsing instead of making chat templates, this forces us to use community written chat templates instead, see https://github.com/ggml-org/llama.cpp/pull/14862#discussion_r2229652708

We understand the difficulty, however we don't plan to release official ones. We have several advantages doing it via `mistral-common`: validation, aggregation, testing, ... And releasing chat template would add more burden to releases + maintaining it.

This is why we're trying to find ways to ease the usage of `mistral-common` and seek feedback on how it can be improved.

ngxson commented on convert_hf_to_gguf.py:
```diff
@@ -7581,6 +7682,132 @@ def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iter
         return [(self.map_tensor_name(name), data_torch)]
 
 
+class MistralModel(TextModel):
```
Maybe doing this will reduce the function to just `set_gguf_parameters` ?

```suggestion
class MistralModel(LlamaModel):
```

juliendenize replied:
Done

ngxson commented on convert_hf_to_gguf.py:
```diff
@@ -7581,6 +7682,132 @@ def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iter
+        return [(self.map_tensor_name(name), data_torch)]
+
+
+class PixtralModel(MmprojModel):
```
Same here, probably this is shorter

```suggestion
class PixtralModel(LlavaVisionModel):
```

juliendenize replied:
Done

ngxson commented on convert_hf_to_gguf.py:
```diff
@@ -7730,6 +7957,16 @@ def parse_args() -> argparse.Namespace:
+        help="Whether the model is stored following the Mistral format.",
+    )
+    parser.add_argument(
+        "--n-ctx",
```
Could we maybe have a list of predefined mapping `model --> n_ctx` instead?

Having this as a required param is quite confusing tbh. Ideally it should be inside `params.json`, otherwise we should make it such that users are not required to enter it manually.

ngxson commented on gguf-py/gguf/tensor_mapping.py:
```diff
@@ -1059,6 +1059,8 @@ class TensorNameMap:
             "multi_modal_projector.linear_{bid}",
             "visual.merger.mlp.{bid}", # qwen2vl
+            "vision_language_adapter.w_in", # pixtral
+            "vision_language_adapter.w_out", # pixtral
```
this is potentially be wrong, as we expect a `{bid}` number in the name. Otherwise, you should have a custom renaming from `w_in` to `multi_modal_projector.linear_1` inside the `modify_tensors` function. Same for `w_out`

ngxson commented on convert_hf_to_gguf.py:
```diff
@@ -7581,6 +7682,132 @@ def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iter
+        ):
+            # process vision tensors
+            if name.endswith(("q_proj.weight", "q_proj.bias")):
+                data_torch = MistralModel.permute(data_torch, n_head, n_head)
```
IIRC the original model is not permuted, only the HF model is.

So having this line could lead to a broken model

ngxson reviewed (commented):
Please note that when I merge https://github.com/ggml-org/llama.cpp/pull/14862, it will create some conflicts in this PR. I will make a PR to your repo to resolve the conflicts.

In the meantime, I think it's important to test the output from your script. As mentioned in some comments, this probably not yet working.

juliendenize committed:
refactor

juliendenize commented:
> Please note that when I merge #14862, it will create some conflicts in this PR. I will make a PR to your repo to resolve the conflicts.
> 
> In the meantime, I think it's important to test the output from your script. As mentioned in some comments, this probably not yet working.

thanks for the review, I updated the PR. Tested on Mistral-Small-3.2 with mmproj. I used the jinja template from 3.1.

```python
from datetime import datetime, timedelta

from openai import OpenAI
from huggingface_hub import hf_hub_download

# Modify OpenAI's API key and API base to use vLLM's API server.
openai_api_key = "EMPTY"
openai_api_base = "http://localhost:8080/v1"

TEMP = 0.15
MAX_TOK = 131072

client = OpenAI(
    api_key=openai_api_key,
    base_url=openai_api_base,
)

models = client.models.list()
model = models.data[0].id


def load_system_prompt(repo_id: str, filename: str) -> str:
    file_path = hf_hub_download(repo_id=repo_id, filename=filename)
    with open(file_path, "r") as file:
        system_prompt = file.read()
    today = datetime.today().strftime("%Y-%m-%d")
    yesterday = (datetime.today() - timedelta(days=1)).strftime("%Y-%m-%d")
    model_name = repo_id.split("/")[-1]
    return system_prompt.format(name=model_name, today=today, yesterday=yesterday)


model_id = "mistralai/Mistral-Small-3.2-24B-Instruct-2506"
SYSTEM_PROMPT = load_system_prompt(model_id, "SYSTEM_PROMPT.txt")
image_url = "https://cdn.lospec.com/thumbnails/gallery/gabigaie/a-little-pikachu-default.png"

messages = [
    {"role": "system", "content": SYSTEM_PROMPT},
    {
        "role": "user",
        "content": [
            {
                "type": "text",
                "text": "What is it and which side is the tail ?",
            },
            {"type": "image_url", "image_url": {"url": image_url}},
        ],
    },
]


response = client.chat.completions.create(
    model=model,
    messages=messages,
    temperature=TEMP,
    max_tokens=MAX_TOK,
)

print(response.choices[0].message.content)
```

Based on your second paragraph i didn't rebase but lmk if you want me to do it.

ngxson commented:
Thanks, it looks better now.

> Based on your second paragraph i didn't rebase but lmk if you want me to do it.

Yes please do a rebase. I thought about doing it myself but turns out I still haven't had time.

juliendenize committed:
revert

juliendenize committed:
remove duplicate

juliendenize committed:
Remove duplication code

juliendenize committed:
Fixes

juliendenize committed:
Fix flake issues

juliendenize committed:
Apply comments

juliendenize committed:
Apply comments

juliendenize commented on convert_hf_to_gguf.py:
```diff
@@ -1951,44 +2035,7 @@ def set_vocab(self):
-        self.gguf_writer.add_vocab_size(vocab.vocab_size)
-
-        self.gguf_writer.add_add_bos_token(True)
-        self.gguf_writer.add_add_eos_token(False)
+        self._set_vocab_mistral()
```
Not sure in `set_vocab_tekken` you also have the following:
```python
script_dir = Path(__file__).parent
        template_path = script_dir / "models/templates/unsloth-mistral-Devstral-Small-2507.jinja"
        with open(template_path, "r", encoding="utf-8") as f:
            template = f.read()
            self.gguf_writer.add_chat_template(template)
```

LMK if you still want to discard this.

juliendenize commented on convert_hf_to_gguf.py:
```diff
@@ -1951,44 +2035,7 @@ def set_vocab(self):
-        self.gguf_writer.add_add_eos_token(False)
+        self._set_vocab_mistral()
 
         script_dir = Path(__file__).parent
```
Ok now got it ignore previous comment.

juliendenize committed:
Apply comments

juliendenize committed:
Fix remote

juliendenize commented on convert_hf_to_gguf.py:
```diff
@@ -7737,6 +7813,44 @@ def prepare_tensors(self):
+    model_name = "Mistral"
+    hf_arch = ""
+    is_mistral_format = True
+    undo_permute = True
```
Changed this, weirdly enough it was working before and now still.

But yeah checked the conversion scrip mistral -> hf we do permutation so we shouldn't have to undo it here.

juliendenize commented on gguf-py/gguf/utility.py:
```diff
@@ -120,9 +120,13 @@ def get_list_tensors_hf_model(cls, model_id: str) -> dict[str, RemoteTensor]:
             return cls.get_list_tensors(url)
+        if is_single_file_consolidated:
+            url = f"{cls.BASE_DOMAIN}/{model_id}/resolve/main/consolidated.safetensors"
+            return cls.get_list_tensors(url)
```
Reverted the change for SafetensorRemote.

I updated the function get_remote_tensors to properly handle the mistral format.

ngxson commented on convert_hf_to_gguf.py:
```diff
@@ -1990,16 +1993,59 @@ def set_vocab_tekken(self):
+        if path_tekken_json.is_file() and not path_tokenizer_json.is_file():
+            self._set_vocab_mistral()
+
+            script_dir = Path(__file__).parent
+            template_path = script_dir / "models/templates/unsloth-mistral-Devstral-Small-2507.jinja"
+            with open(template_path, "r", encoding="utf-8") as f:
+                template = f.read()
+                self.gguf_writer.add_chat_template(template)
```
I'm wondering if we can also move this to `_set_vocab_mistral()`, as for now GGUF converted via `--mistral-format` won't have any templates built-in, thus it won't be usable out-of-the box with tools based on llama.cpp.

IMO forcing people to use both python (to format the chat) and llama.cpp at the same time may not be a good user experience, so having built-in jinja template should still be a requirement.

Also self-note: one thing missing from this code, we should make sure we are using the correct tekken version for the given model, maybe as a `assert tekken_json["config"]["version"] == "v7"`. But I can add it later.

ngxson commented on convert_hf_to_gguf.py:
```diff
@@ -104,18 +109,26 @@ def __init__(self, dir_model: Path, ftype: gguf.LlamaFileType, fname_out: Path,
                 logger.info(f"Using remote model with HuggingFace id: {remote_hf_model_id}")
-                remote_tensors = gguf.utility.SafetensorRemote.get_list_tensors_hf_model(remote_hf_model_id)
+
+                if not self.is_mistral_format:
+                    remote_tensors = gguf.utility.SafetensorRemote.get_list_tensors_hf_model(remote_hf_model_id)
+
+                else:
+                    url = f"{gguf.utility.SafetensorRemote.BASE_DOMAIN}/{remote_hf_model_id}/resolve/main/consolidated.safetensors"
+                    remote_tensors = gguf.utility.SafetensorRemote.get_list_tensors(url)
+                
                 self.tensor_names = set(name for name in remote_tensors.keys())
-                for name, remote_tensor in gguf.utility.SafetensorRemote.get_list_tensors_hf_model(remote_hf_model_id).items():
+                for name, remote_tensor in remote_tensors.items():
                     yield (name, LazyTorchTensor.from_remote_tensor(remote_tensor))
 
             self.get_tensors = get_remote_tensors
```
As discussed about remote_safetensors, I think these changes should also be reverted

ngxson reviewed (approved):
Looks good overall, nice contribution!

We can merge once the 2 pending comments are all resolved.

juliendenize committed:
add default chat template

juliendenize committed:
Revert

juliendenize committed:
nit

juliendenize reviewed (commented):
Think I answered your comments.

Regarding the remote I think it would have been nice to allow remote download for mistral format but I reverted as requested.

Regarding chat templates, i created a method for it that should be easily expandable.

juliendenize commented:
@CISC @ngxson gentle ping to know if there is anything more I can do to help merging this one.

Son alerted me that he'll be busy, so I understand if for now this PR is on hold. Just so you know, next week I'll be on vacations !

broadbit-hu commented:
> ```
>             case LLAMA_VOCAB_PRE_TYPE_TEKKEN:
>                 // original regex from tokenizer.json
>                 // "[^\\r\\n\\p{L}\\p{N}]?[\\p{Lu}\\p{Lt}\\p{Lm}\\p{Lo}\\p{M}]*[\\p{Ll}\\p{Lm}\\p{Lo}\\p{M}]+|[^\\r\\n\\p{L}\\p{N}]?[\\p{Lu}\\p{Lt}\\p{Lm}\\p{Lo}\\p{M}]+[\\p{Ll}\\p{Lm}\\p{Lo}\\p{M}]*|\\p{N}| ?[^\\s\\p{L}\\p{N}]+[\\r\\n/]*|\\s*[\\r\\n]+|\\s+(?!\\S)|\\s+"
>                 regex_exprs = {
>                     "[^\\r\\n\\p{L}\\p{N}]?((?=[\\p{L}])([^a-z]))*((?=[\\p{L}])([^A-Z]))+|[^\\r\\n\\p{L}\\p{N}]?((?=[\\p{L}])([^a-z]))+((?=[\\p{L}])([^A-Z]))*|\\p{N}| ?[^\\s\\p{L}\\p{N}]+[\\r\\n/]*|\\s*[\\r\\n]+|\\s+(?!\\S)|\\s+",
>                 };
>                 break;
> ```

@juliendenize What do you think about these differences in the Tekken tokenizer that affect Unicode characters?

juliendenize commented:
@broadbit-hu 
Not sure to understand the question, I think it is handled [here](https://github.com/ggml-org/llama.cpp/pull/14737/files#diff-ec77d8003b92ff283179456d36b8b56abf635e7b1232e70daf16676e8920ccf1R1972-R1974) if this was the request.

If the request is to improve, it is out of scope rn (at least to me), not sure there was requests to improve or performance issues raised.

ngxson reviewed (approved):
@CISC please merge when you're ready

juliendenize commented:
Jusdt got back from vaca, thanks for your help merging this one @CISC @ngxson !
